mlflow/server/auth/__init__.py

```diff
@@ -414,20 +418,52 @@ def _get_proxy_artifact_validator(
     }.get(method)
 
 
-@catch_mlflow_exception
-def _before_request():
-    if is_unprotected_route(request.path):
-        return
+def authenticate_request() -> Union[Authorization, Response]:
+    """Use configured authorization function to get request authorization."""
+    auth_func = get_auth_func(auth_config.authorization_function)
+    return auth_func()
+
+
+@functools.lru_cache(maxsize=None)
+def get_auth_func(authorization_function: str) -> Callable[[], Union[Authorization, Response]]:
+    """Import and return the specified authorization function.
+
+    :param authorization_function: A string of the form "module.submodule:auth_func
+    """
+    mod_name, fn_name = authorization_function.split(":", 1)
+    module = importlib.import_module(mod_name)
+    return getattr(module, fn_name)
+
 
+def authenticate_request_basic_auth() -> Union[Authorization, Response]:
+    """Authenticate the request using basic auth."""
     if request.authorization is None:
         return make_basic_auth_response()
 
     username = request.authorization.username
     password = request.authorization.password
-    if not store.authenticate_user(username, password):
+    if store.authenticate_user(username, password):
+        return request.authorization
+    else:
         # let user attempt login again
         return make_basic_auth_response()
 
+
+@catch_mlflow_exception
+def _before_request():
+    if is_unprotected_route(request.path):
+        return
+
+    authorization = authenticate_request()
+    if isinstance(authorization, Response):
+        return authorization
+    elif not isinstance(authorization, Authorization):
+        raise MlflowException(
+            f"Unsupported result type from {auth_config.authorization_function}: "
+            f"'{type(authorization).__name__}'",
+            INTERNAL_ERROR,
+        )
```

Comment:
```suggestion
    if isinstance(authorization, (Response, Authorization)):
        return authorization
    else:
        raise MlflowException(
            f"Unsupported result type from {auth_config.authorization_function}: "
            f"'{type(authorization).__name__}'",
            INTERNAL_ERROR,
        )
```

Is this more straightforward?

Reply:
That change looks like it introduces a bug. If it's an `Authorization` instance, we don't want to return immediately. We want to continue executing the remaining lines in the function:
```
    # admins don't need to be authorized
    if sender_is_admin():
        return

    # authorization
    if validator := BEFORE_REQUEST_VALIDATORS.get((request.path, request.method)):
        if not validator():
            return make_forbidden_response()
    elif _is_proxy_artifact_path(request.path):
        if validator := _get_proxy_artifact_validator(request.method, request.view_args):
            if not validator():
                return make_forbidden_response()
```

Reply:
ah I see, the current code returns None if `authorization` is `Authorization`.